From 28b7fca62cd46c7d0c8282c3acef5c5389041f69 Mon Sep 17 00:00:00 2001
From: Noam Lessner <nlessner@habana.ai>
Date: Thu, 03 Feb 2022 11:21:17 +0200
Subject: [PATCH] [SW-56024] fwupd: fix offset adjust

fix adjusting offset only if there is a response ready to be picked up
by the fwl. This is to fix offsetting and masking the value when 0
causing the offset to be 0x7000000 insted of 0.

Signed-off-by: Noam Lessner <nlessner@habana.ai>
Change-Id: Iec5fc107370c5533202fa518a3c090d9e7f1470b
---

diff --git a/lib/preboot/module_fwupd.c b/lib/preboot/module_fwupd.c
index 653d43f..64043ff 100644
--- a/lib/preboot/module_fwupd.c
+++ b/lib/preboot/module_fwupd.c
@@ -167,7 +167,15 @@
 
 	status.val = 0;
 	status.status = sts;
-	status.offset = (u32)((address - SDRAM_BASE) & ~HL_UPD_STS_OFFSET_MASK);
+
+	/* only if we get a address we need to calc offset */
+	if (address >=  SDRAM_BASE)
+		status.offset = (u32)((address - SDRAM_BASE) &
+				~HL_UPD_STS_OFFSET_MASK);
+	else
+		status.offset = (u32)(address & ~HL_UPD_STS_OFFSET_MASK);
+
+
 	state->sts_writel(status.val);
 }
 
